src/QsFmt/Formatter/Formatter.fs

```diff
@@ -33,14 +33,24 @@ let parse (source: string) =
 
 [<CompiledName "Format">]
 let format source =
-    parse source
-    |> Result.map (
-        curry collapsedSpaces.Document ()
-        >> curry operatorSpacing.Document ()
-        >> curry newLines.Document ()
-        >> curry indentation.Document 0
-        >> printer.Document
-    )
+    let formatDocument document =
+        // Test whether there is data loss during parsing and unparsing
+        if printer.Document document = source then
+            // The actuall format process
+            document
+            |> curry collapsedSpaces.Document ()
+            |> curry operatorSpacing.Document ()
+            |> curry newLines.Document ()
+            |> curry indentation.Document 0
+            |> printer.Document
+        // Report error if the unparsing result does match the original source
+        else
+            failwith (
+                "The formater does not work properly. "
+                + "Please let us know by filing a new issue in https://github.com/microsoft/qsharp-compiler/issues/new/choose."
```

Comment:
It would be nice if the error message was a little more specific - for example it could say that the formatter's syntax tree is not consistent with the input source code. Also, including the unparsed document in the message could help with debugging.